pallets/ethereum-checked/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+};
+
+pub use pallet::*;
+
+mod mock;
+mod tests;
+
+/// Origin for dispatch-able calls.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub enum RawOrigin<AccountId> {
+    XcmEthereumTx(AccountId),
+}
+
+/// Ensure the origin is with XCM calls.
+pub struct EnsureXcmEthereumTx<AccountId>(PhantomData<AccountId>);
+impl<O: Into<Result<RawOrigin<AccountId>, O>> + From<RawOrigin<AccountId>>, AccountId: Decode>
+    EnsureOrigin<O> for EnsureXcmEthereumTx<AccountId>
+{
+    type Success = AccountId;
+
+    fn try_origin(o: O) -> Result<Self::Success, O> {
+        o.into().map(|o| match o {
+            RawOrigin::XcmEthereumTx(account_id) => account_id,
+        })
+    }
+
+    #[cfg(feature = "runtime-benchmarks")]
+    fn try_successful_origin() -> Result<O, ()> {
+        let zero_account_id =
+            AccountId::decode(&mut TrailingZeroInput::zeroes()).map_err(|_| ())?;
+        Ok(O::from(RawOrigin::XcmEthereumTx(zero_account_id)))
+    }
+}
+
+/// Transaction kind.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub enum CheckedEthereumTxKind {
+    /// The tx is from XCM remote call.
+    Xcm,
+    /// The tx is from cross-VM call.
+    Xvm,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_evm::Config {
+        /// Reserved Xcmp weight for block gas limit calculation.
+        type ReservedXcmpWeight: Get<Weight>;
+
+        /// Xcm transaction weight limit, for block gas limit calculation.
+        type XvmTxWeightLimit: Get<Weight>;
+
+        /// Invalid tx error.
+        type InvalidEvmTransactionError: From<InvalidEvmTransactionError>;
+
+        /// Validated tx execution.
+        type ValidatedTransaction: ValidatedTransaction;
+
+        /// Account mapping.
+        type AccountMapping: AccountMapping<Self::AccountId>;
+
+        /// Origin for `transact` call.
+        type TransactOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+    }
+
+    #[pallet::origin]
+    pub type Origin<T> = RawOrigin<<T as frame_system::Config>::AccountId>;
+
+    /// Global nonce for all transactions to avoid hash collision, which is
+    /// caused by the same dummy signatures for all transactions.
+    #[pallet::storage]
+    pub type Nonce<T: Config> = StorageValue<_, U256, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Transact an Ethereum transaction. Similar to `pallet_ethereum::Transact`,
+        /// but is only for XCM remote call.
+        #[pallet::call_index(0)]
+        #[pallet::weight({
+            let weight_limit = T::GasWeightMapping::gas_to_weight(tx.gas_limit.unique_saturated_into(), false);
+            // `Nonce` storage read 1, write 1.
+            weight_limit.saturating_add(T::DbWeight::get().reads_writes(1, 1))
```

Comment:
Personally, I'd like to see proper benchmark number here. I guess we don't expect the rest of the logic to have significant impact, but it can still be useful to be able to re-run it later.

Reply:
Based on the current gas fee model, `gas_limit` has included all the costs for the transaction, the only part not covered is the preparation work before applying it via `pallet-ethereum`.

The DB read/write result via benchmarks could be misleading IMO, as it includes DB operations within `pallet-ethereum`, which had already been taken into consideration with provided `gas_limit`. We could revisit here when align fees work started.

Reply:
Of course, the `do_transact` function would have to be stripped of `T::ValidatedTransaction::apply(source, tx)` to properly measure it.

Maybe it's not worth doing right now 🤷‍♂️

Reply:
It makes sense. Let me add the benchmarks in the follow-up PR of XVM & Shibuya integration.